### ISSUE TITLE
Building errors - [ Fedora using system libs ]

### ISSUE THREAD

**Issue by Enchufa2** [shikokuchuo/nanonext#5]
I'm having trouble bringing nanonext to [this repository](https://copr.fedorainfracloud.org/coprs/iucar/cran/package/R-CRAN-nanonext/) for Fedora, as you can see from the list of failed builds. Both required libraries are available in Fedora, but I get the following errors:

```
** using staged installation
NANONEXT_SYS flag set
Using 'libnng' and 'libmbedtls' installed in '/usr' ...
** libs
gcc -m64 -I"/usr/include/R" -DNDEBUG   -I/usr/local/include  -I/usr/include  -fvisibility=hidden -fpic  -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -c aio.c -o aio.o
aio.c: In function 'rnng_aio_result':
aio.c:148:7: warning: implicit declaration of function 'nng_aio_busy'; did you mean 'nng_aio_begin'? [-Wimplicit-function-declaration]
  148 |   if (nng_aio_busy(aiop->aio))
      |       ^~~~~~~~~~~~
      |       nng_aio_begin
gcc -m64 -I"/usr/include/R" -DNDEBUG   -I/usr/local/include  -I/usr/include  -fvisibility=hidden -fpic  -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -c core.c -o core.o
gcc -m64 -I"/usr/include/R" -DNDEBUG   -I/usr/local/include  -I/usr/include  -fvisibility=hidden -fpic  -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -c init.c -o init.o
gcc -m64 -I"/usr/include/R" -DNDEBUG   -I/usr/local/include  -I/usr/include  -fvisibility=hidden -fpic  -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -c opts.c -o opts.o
gcc -m64 -I"/usr/include/R" -DNDEBUG   -I/usr/local/include  -I/usr/include  -fvisibility=hidden -fpic  -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -c sha.c -o sha.o
sha.c: In function 'rnng_sha224':
sha.c:78:8: error: void value not ignored as it ought to be
   78 |     xc = mbedtls_sha256(buf, sz, output, 1);
      |        ^
sha.c: In function 'rnng_sha256':
sha.c:159:8: error: void value not ignored as it ought to be
  159 |     xc = mbedtls_sha256(buf, sz, output, 0);
      |        ^
sha.c: In function 'rnng_sha384':
sha.c:240:8: error: void value not ignored as it ought to be
  240 |     xc = mbedtls_sha512(buf, sz, output, 1);
      |        ^
sha.c: In function 'rnng_sha512':
sha.c:321:8: error: void value not ignored as it ought to be
  321 |     xc = mbedtls_sha512(buf, sz, output, 0);
      |        ^
make: *** [/usr/lib64/R/etc/Makeconf:168: sha.o] Error 1
ERROR: compilation failed for package 'nanonext'
* removing '/builddir/build/BUILDROOT/R-CRAN-nanonext-0.5.5-1.fc36.copr4863362.x86_64/usr/local/lib/R/library/nanonext'
error: Bad exit status from /var/tmp/rpm-tmp.XWG99o (%install)
```

You can check the full log [here](https://download.copr.fedorainfracloud.org/results/iucar/cran/fedora-36-x86_64/04863362-R-CRAN-=nanonext/builder-live.log.gz).

On the other hand, it would be great if the package could automatically detect if the system libraries are available under `/usr` (not only `/usr/local`) and use them in such a case (here I patched the configure to use `/usr` instead of `/usr/local`).

**Comment by shikokuchuo** [comment]
Hi, thanks for the message. It is written in the readme - the reason the package downloads the libraries to build from source is it requires a newer version of NNG than that available in repos. Is it possible you allow the package to be built as intended, or let me know if there is anything I can do to help.

**Comment by shikokuchuo** [comment]
That is also the reason it does not pick up libraries under /usr

**Comment by Enchufa2** [comment]
Oh, I see. Anyway, I think it's still valuable if the configure script checks if the system libraries are available. You could check the version by e.g. checking certain functions, or the soname version. Then, the installation would just work when we update the libraries in Fedora.

**Comment by shikokuchuo** [comment]
I agree with you - that was indeed the approach taken previously. And will revert to that as soon as the author of NNG issues a release version that will eventually get to system repos. It is just designed this way so it is ensured to work at the moment. Thanks!

**Comment by shikokuchuo** [comment]
This is fully addressed in commit 1a840c1 - package now bundles NNG source and uses system MbedTLS where available.

**Comment by Enchufa2** [comment]
Since v0.6.0 this works fine and fast in Fedora:

```bash
> install.packages("nanonext")
Install system packages...
R-CRAN-nanonext-0.7.3-1.fc37.copr5283292.x86_64 620 kB/s | 1.3 MB     00:02    
  Preparing        :                                                        1/1 
  Installing       : R-CRAN-nanonext-0.7.3-1.fc37.copr5283292.x86_64        1/1 
  Running scriptlet: R-CRAN-nanonext-0.7.3-1.fc37.copr5283292.x86_64        1/1 
  Verifying        : R-CRAN-nanonext-0.7.3-1.fc37.copr5283292.x86_64        1/1
```

and mbedtls is automagically pulled too if not present already (in the case above it was already present).
